src/backend/distributed/metadata/metadata_utility.c

```diff
@@ -2357,3 +2363,1594 @@ IsForeignTable(Oid relationId)
+
+			pfree(task);
+			task = NULL;
+		}
+
+		systable_endscan(scanDescriptor);
+	}
+
+	table_close(pgDistBackgroundTasks, NoLock);
+
+	return task;
+}
+
+
+/*
+ * GetBackgroundJobByJobId loads a BackgroundJob from the catalog into memory. Return's a
+ * null pointer if no job exist with the given JobId.
+ */
+BackgroundJob *
+GetBackgroundJobByJobId(int64 jobId)
+{
+	ScanKeyData scanKey[1] = { 0 };
+	bool indexOK = true;
+
+	Relation pgDistBackgroundJobs =
+		table_open(DistBackgroundJobsRelationId(), AccessShareLock);
+
+	/* pg_dist_background_tasks.job_id == $jobId */
+	ScanKeyInit(&scanKey[0], Anum_pg_dist_background_jobs_job_id,
+				BTEqualStrategyNumber, F_INT8EQ, Int64GetDatum(jobId));
+
+	SysScanDesc scanDescriptor =
+		systable_beginscan(pgDistBackgroundJobs, DistBackgroundJobsJobIdIndexId(),
+						   indexOK, NULL, lengthof(scanKey), scanKey);
+
+	HeapTuple taskTuple = systable_getnext(scanDescriptor);
+	BackgroundJob *job = NULL;
+	if (HeapTupleIsValid(taskTuple))
+	{
+		TupleDesc tupleDescriptor = RelationGetDescr(pgDistBackgroundJobs);
+		job = DeformBackgroundJobHeapTuple(tupleDescriptor, taskTuple);
+	}
+
+	systable_endscan(scanDescriptor);
+	table_close(pgDistBackgroundJobs, AccessShareLock);
+
+	return job;
+}
+
+
+/*
+ * GetBackgroundTaskByTaskId loads a BackgroundTask from the catalog into memory. Return's
+ * a null pointer if no job exist with the given JobId and TaskId.
+ */
+BackgroundTask *
+GetBackgroundTaskByTaskId(int64 jobId, int64 taskId)
+{
+	ScanKeyData scanKey[2] = { 0 };
+	bool indexOK = true;
+
+	Relation pgDistBackgroundTasks =
+		table_open(DistBackgroundTasksRelationId(), AccessShareLock);
+
+	/* pg_dist_background_tasks.job_id == $jobId */
+	ScanKeyInit(&scanKey[0], Anum_pg_dist_background_tasks_job_id,
+				BTEqualStrategyNumber, F_INT8EQ, Int64GetDatum(jobId));
+
+	/* pg_dist_background_tasks.task_id == $taskId */
+	ScanKeyInit(&scanKey[1], Anum_pg_dist_background_tasks_task_id,
+				BTEqualStrategyNumber, F_INT8EQ, Int64GetDatum(taskId));
+
+	SysScanDesc scanDescriptor =
+		systable_beginscan(pgDistBackgroundTasks,
+						   DistBackgroundTasksTaskIdIndexId(),
+						   indexOK, NULL, lengthof(scanKey), scanKey);
+
+	HeapTuple taskTuple = systable_getnext(scanDescriptor);
+	BackgroundTask *task = NULL;
+	if (HeapTupleIsValid(taskTuple))
+	{
+		TupleDesc tupleDescriptor = RelationGetDescr(pgDistBackgroundTasks);
+		task = DeformBackgroundTaskHeapTuple(tupleDescriptor, taskTuple);
+	}
+
+	systable_endscan(scanDescriptor);
+	table_close(pgDistBackgroundTasks, AccessShareLock);
+
+	return task;
+}
+
+
+/*
+ * JobTasksStatusCount scans all tasks associated with the provided job and count's the
+ * number of tasks that are tracked in each state. Effectively grouping and counting the
+ * tasks by their state.
+ */
+static bool
+JobTasksStatusCount(int64 jobId, int *blocked, int *runnable, int *running, int *done,
+					int *error, int *unscheduled, int *cancelled)
+{
+#define CLEAR_OPTIONAL_COUNTER(counter) \
```

Comment:
what if we made them non-optional?  seems this guy only has a single caller that sets all of them

Reply:
Yeah, will do, this is one of those functions that I am supper unhappy about, but couldn't quickly come up with a different approach compared to a manual group by count(*).

Since there is only one use that passes them all it makes sense to make them non-optional. Was even thinking of simply returning a struct by value instead of the long list of counts. It started with one or two counts :P, never revisited.